airflow/serialization/serialized_objects.py

```diff
@@ -845,18 +845,18 @@ def _serialize_params_dict(cls, params: ParamsDict | dict) -> list[tuple[str, di
         """Serialize Params dict for a DAG or task as a list of tuples to ensure ordering."""
         serialized_params = []
         for k, v in params.items():
-            # TODO: As of now, we would allow serialization of params which are of type Param only.
+            if isinstance(params, ParamsDict):
+                # Use native param object, not resolved value if possible
+                v = params.get_param(k)
             try:
                 class_identity = f"{v.__module__}.{v.__class__.__name__}"
             except AttributeError:
                 class_identity = ""
             if class_identity == "airflow.models.param.Param":
                 serialized_params.append((k, cls._serialize_param(v)))
             else:
-                raise ValueError(
-                    f"Params to a DAG or a Task can be only of type airflow.models.param.Param, "
-                    f"but param {k!r} is {v.__class__}"
-                )
+                # Auto-boy other values into Params object like it is done by DAG parsing as well
```

Comment:
`Auto-boy`?

Reply:
Same question :)

Reply:
Auto-boy... Auto-boy... you don't know Auto-boy? Hahaha... typo!
```suggestion
                # Auto-box other values into Params object like it is done by DAG parsing as well
```